File: step14/main/dao/managerDao.js
"use strict"

module.exports = {
  setConnection(conn) {
    this.connection = conn // 변수를 알아서 만들어준다.
  },

  selectList(pageNo, successFn, errorFn) { // 성공했을 때 호출될 함수, 실패했을 때 호출될 함수
    // alert(currPageNo + 1) // 0번째 태그의 콘텐츠만 가져옴, 콘텐츠 = 시작태그와 끝태그 사이에 있는 값
    this.connection.query(
      'select m.mno, m.name, m.tel, m.email, r.posi \
      from mgr r inner join memb m on r.mrno = m.mno \
      order by m.mno asc\
      limit ?, ?',
      [(pageNo -1) * 3, 3],
      function (error, results) {
        if (error) {
          errorFn(error)
        } else {
          successFn(results)
        }
    }) // connection.query
  }, // listStudent

  countAll(successFn, errorFn) {
    this.connection.query(
      'select count(*) cnt from mgr',
      function (error, results) {
        if (error) {
          errorFn(error)
        } else {
          successFn(results)
        }
    }) // connection.query
  }, // countAllStudent

  selectOne(no, successFn, errorFn){
    this.connection.query(
      'select m.mno, m.name, m.tel, m.email, r.posi, r.fax, r.path \
      from mgr r inner join memb m on r.mrno = m.mno \
      where r.mrno = ?',
      [no],
      function(error, results) {
        if (error) {
          errorFn(error)
        } else {
          successFn(results[0]) // select가 한개라서 0번으로 호출
        }
      }) // connection.query
  }, // selectOneStudent

  insert(manager, successFn, errorFn) {
    this.connection.query(
      'insert into mgr(mrno, posi, fax, path) values(?, ?, ?, ?)',
      [manager.no, manager.posi, manager.fax, manager.path],
      function(error, results) {
        if (error) {
          errorFn(error)
        } else {
          successFn(results)
        }
    }) // connection.query
  }, //insertStudent

  update(manager, successFn, errorFn) {
    this.connection.query(
      'update mgr set posi=?, fax=?, path=? where mrno=?',
      [manager.posi, manager.fax, manager.path, manager.no],
      function(error, results) {
        if (error) {
          errorFn(error)
        } else {
          successFn(results)
        }
    }) // connection.query
  }, // updateStudent

  delete(no, successFn, errorFn) {
    this.connection.query(
      'delete from mgr where mrno=?',
      [no],
      function(error, results) {
        if (error) {
          errorFn(error)
        } else {
          successFn(results)
        }
    }) // connection.query
  } // deleteStudent
} // module











//

File: web01/src/step03/ManagerDao.java
package step03;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;

public class ManagerDao {
  DBConnectionPool conPool;

  public ManagerDao(DBConnectionPool conPool) {
    this.conPool = conPool;
  }

  public List<Manager> selectList() throws Exception {
    // 사용할 커넥션을 DBConnectionPool로부터 빌린다.
    Connection con = conPool.getConnection();

    try (
        PreparedStatement stmt = con.prepareStatement(
            "select mr.mrno, m.name"
            + " from mgr mr inner join memb m on mr.mrno=m.mno"
            + " order by m.name asc");) {


      ArrayList<Manager> list = new ArrayList<>();
      try (ResultSet rs = stmt.executeQuery();) {
        Manager manager = null;
        while (rs.next()) {
          manager = new Manager();
          manager.setNo(rs.getInt("mrno"));
          manager.setName(rs.getString("name"));

          list.add(manager);
        }
      }
      return list;

    } finally { // 다 쓴 커넥션을 반납하기 위해서
      // finally 블록은 try 블록을 벗어나기 전에 반드시 실행되는 블록이다.
      // try 블록에서 return 문을 실행하기 전에 이 블록을 실행한다.
      conPool.returnConnection(con);
    }
  }

  public Manager selectOne(int no) throws Exception {
    Connection con = conPool.getConnection();

    try (
        PreparedStatement stmt = con.prepareStatement(
            "select mr.mrno, m.name"
            + " from mgr mr inner join memb m on mr.mrno=m.mno"
            + " order by m.name asc");) {

      stmt.setInt(1, no);

      try (ResultSet rs = stmt.executeQuery();) {
        if (!rs.next()) {
          return null;
        }

        Manager manager = new Manager();
        manager.setNo(rs.getInt("mrno"));
        manager.setName(rs.getString("name"));
        return manager;
      }

    } finally { // 다 쓴 커넥션을 반납하기 위해서
      // finally 블록은 try 블록을 벗어나기 전에 반드시 실행되는 블록이다.
      // try 블록에서 return 문을 실행하기 전에 이 블록을 실행한다.
      conPool.returnConnection(con);
    }
  }

  public int insert(Member member) throws Exception {
    Connection con = conPool.getConnection();
    try (
        PreparedStatement stmt = con.prepareStatement(
            "insert into memb(name, tel, email, pwd) value(?,?,?, password(?))");
        ) {
      stmt.setString(1, member.getName());
      stmt.setString(2, member.getTel());
      stmt.setString(3, member.getEmail());
      stmt.setString(4, member.getPassword());
      return stmt.executeUpdate();
    }finally {
      conPool.returnConnection(con);
    }
  }

  public int delete(int no) throws Exception {
    Connection con = conPool.getConnection();
    try (
        PreparedStatement stmt = con.prepareStatement(
            "delete from memb where mno=?");
        ) {

      stmt.setInt(1, no);
      return stmt.executeUpdate();

    } finally {
      conPool.returnConnection(con);
    }
  }

  public int update(Member member) throws Exception {
    Connection con = conPool.getConnection();
    try (
        PreparedStatement stmt = con.prepareStatement(
            "update memb set name=?, tel=?, email=?, pwd=password(?) where mno=?");
        ) {
      stmt.setString(1, member.getName());
      stmt.setString(2, member.getTel());
      stmt.setString(3, member.getEmail());
      stmt.setString(4, member.getPassword());
      stmt.setInt(5, member.getNo());
      return stmt.executeUpdate();
    } finally {
      conPool.returnConnection(con);
    }
  }
}
